In windows can't use arrow navigation exec bash into container

BUG REPORT INFORMATION


**docker version:**
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): windows/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64

**docker info:**
Containers: 15
Images: 309
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 339
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 4.0.3-boot2docker
Operating System: Boot2Docker 1.6.2 (TCL 5.4); master : 4534e65 - Wed May 13 21:24:28 UTC 2015
CPUs: 4
Total Memory: 1.957 GiB
Name: boot2docker
ID: FKPI:HE56:LNVN:B25Q:RDMA:YB5E:AENE:Y2HR:72C4:CNUS:LNMP:3G7M
Debug mode (server): true
Debug mode (client): false
Fds: 32
Goroutines: 49
System Time: Tue Jun  9 06:00:31 UTC 2015
EventsListeners: 0
Init SHA1: 7f9c6798b022e64f04d2aff8c75cbf38a2779493
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Username: rasterdocker
Registry: [https://index.docker.io/v1/]

**uname -a:**
MINGW64 _NT-6.1 raster-dev-PC 2.1.2(0.288/5/3) 2015-05-25 15:48 x86_64 Msys


Provide additional environment details (AWS, VirtualBox, physical, etc.):
Windows 7 using native windows client in either cmd or msys2.

List the steps to reproduce the issue:
1)  docker run --rm -it --entrypoint=//bin/bash ubuntu:14.04
2) type anything, press enter, type some more
3) Hit left, right, up, down, home, end.

Describe the results you received:
Nothing happens: cursor does not move.

Describe the results you expected:
cursor moves left, right, up one in history, down one in history, beginning of line, end of line

Provide additional info you think is important:
Tried it on 2 windows 7 machines, same results.
It does work if I do boot2docker ssh first and follow same steps.

I tried that but it just ended up hitting an invalid memory address or nil pointer dereference in the log routine.

``Administrator@WIN-9QQMSEQFG8N E:\go\src\github.com\docker\docker\docker         
$ docker run -it --rm windowsservercore cmd                                     
panic: runtime error: invalid memory address or nil pointer dereference         
[signal 0xc0000005 code=0x0 addr=0x28 pc=0x48b016]                              
                                                                                
goroutine 18 [running]:                                                         
github.com/Sirupsen/logrus.(*Logger).Debugf(0x0, 0x1202bd0, 0x4e, 0xc0824d3b20, 
0x3, 0x3)                                                                       
        e:/go/src/github.com/docker/docker/vendor/src/github.com/Sirupsen/logrus
/logger.go:68 +0x26                                                             
github.com/docker/docker/pkg/term/windows.readInputEvents(0x8, 0x8000, 0x0, 0x0,
 0x0, 0x0, 0x0)                                                                 
        e:/go/src/github.com/docker/docker/pkg/term/windows/ansi_reader.go:114 +
0x22f                                                                           
github.com/docker/docker/pkg/term/windows.(*ansiReader).Read(0xc08200c720, 0xc08
24de000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)                                      
        e:/go/src/github.com/docker/docker/pkg/term/windows/ansi_reader.go:73 +0
x437                                                                            
io.Copy(0x4537e98, 0xc0824c3a70, 0x4537e20, 0xc08200c720, 0x0, 0x0, 0x0)        
        C:/Go/src/io/io.go:362 +0x1fd                                           
net.genericReadFrom(0x4537a38, 0xc08204c5b0, 0x4537e20, 0xc08200c720, 0x0, 0x0, 
0x0)                                                                            
        C:/Go/src/net/net.go:404 +0xbc                                          
net.(*TCPConn).ReadFrom(0xc08204c5b0, 0x4537e20, 0xc08200c720, 0x4537e70, 0x0, 0
x0)                                                                             
        C:/Go/src/net/tcpsock_posix.go:72 +0xe7                                 
io.Copy(0x4537a38, 0xc08204c5b0, 0x4537e20, 0xc08200c720, 0x0, 0x0, 0x0)        
        C:/Go/src/io/io.go:358 +0x144                                           
github.com/docker/docker/api/client.func┬╖013()                                 
        e:/go/src/github.com/docker/docker/api/client/hijack.go:228 +0xda       
created by github.com/docker/docker/api/client.(*DockerCli).hijackWithContentTyp
e                                                                               
        e:/go/src/github.com/docker/docker/api/client/hijack.go:240 +0xe00      
                                                                                ``

I found it too ^^', but [binaries from here](https://master.dockerproject.org/) are API 1.22.
But,.... not working :(
This is the trace, doing `docker exec -ti <container> bash`:
````
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x28 pc=0x4a01b3]

goroutine 14 [running]:
github.com/Sirupsen/logrus.(*Logger).Debugf(0x0, 0xaeb080, 0x4e, 0xc082191998, 0x3, 0x3)
	T:/gopath/src/github.com/docker/docker/vendor/src/github.com/Sirupsen/logrus/logger.go:68 +0x23
github.com/docker/docker/pkg/term/windows.readInputEvents(0x3, 0x8000, 0x0, 0x0, 0x0, 0x0, 0x0)
	T:/gopath/src/github.com/docker/docker/pkg/term/windows/ansi_reader.go:114 +0x262
github.com/docker/docker/pkg/term/windows.(*ansiReader).Read(0xc08200d1a0, 0xc082192000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	T:/gopath/src/github.com/docker/docker/pkg/term/windows/ansi_reader.go:73 +0x47b
io.copyBuffer(0x12762f8, 0xc082157be0, 0x1276280, 0xc08200d1a0, 0xc082192000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	T:/Go/src/io/io.go:381 +0x24e
io.Copy(0x12762f8, 0xc082157be0, 0x1276280, 0xc08200d1a0, 0xc082157be0, 0x0, 0x0)
	T:/Go/src/io/io.go:351 +0x6b
net.genericReadFrom(0x1276128, 0xc0820241b0, 0x1276280, 0xc08200d1a0, 0x0, 0x0, 0x0)
	T:/Go/src/net/net.go:559 +0xb8
net.(*TCPConn).ReadFrom(0xc0820241b0, 0x1276280, 0xc08200d1a0, 0xc082191e28, 0x0, 0x0)
	T:/Go/src/net/tcpsock_posix.go:63 +0x2f8
io.copyBuffer(0x1276128, 0xc0820241b0, 0x1276280, 0xc08200d1a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	T:/Go/src/io/io.go:375 +0x187
io.Copy(0x1276128, 0xc0820241b0, 0x1276280, 0xc08200d1a0, 0xc08200d1a0, 0x0, 0x0)
	T:/Go/src/io/io.go:351 +0x6b
github.com/docker/docker/api/client.(*DockerCli).hijackWithContentType.func3(0x1274c00, 0xc08200d1a0, 0x1275048, 0xc0820241b0, 0xc08200db60)
	T:/gopath/src/github.com/docker/docker/api/client/hijack.go:228 +0xe2
created by github.com/docker/docker/api/client.(*DockerCli).hijackWithContentType
	T:/gopath/src/github.com/docker/docker/api/client/hijack.go:240 +0xd95

goroutine 1 [runnable]:
github.com/docker/docker/api/client.(*DockerCli).CmdExec(0xc082072300, 0xc08200e250, 0x3, 0x3, 0x0, 0x0)
	T:/gopath/src/github.com/docker/docker/api/client/exec.go:99 +0xf62
reflect.callMethod(0xc082160300, 0xc082101c78)
	T:/Go/src/reflect/value.go:628 +0x203
reflect.methodValueCall(0xc08200e250, 0x3, 0x3, 0x1, 0xc082160300, 0x0, 0x0, 0xc082160300, 0x0, 0x466a1b, ...)
	T:/Go/src/reflect/asm_amd64.s:29 +0x3d
github.com/docker/docker/cli.(*Cli).Run(0xc082105f20, 0xc08200e240, 0x4, 0x4, 0x0, 0x0)
	T:/gopath/src/github.com/docker/docker/cli/cli.go:89 +0x38a
main.main()
	T:/gopath/src/github.com/docker/docker/docker/docker.go:65 +0x31e

goroutine 5 [syscall]:
os/signal.loop()
	T:/Go/src/os/signal/signal_unix.go:22 +0x1f
created by os/signal.init.1
	T:/Go/src/os/signal/signal_unix.go:28 +0x3e

goroutine 11 [select]:
github.com/docker/docker/api/client.(*DockerCli).hijackWithContentType(0xc082072300, 0xa22618, 0x4, 0xc08200f9a0, 0x4c, 0xa7ef50, 0x10, 0xc08200f901, 0x1274c00, 0xc08200d1a0, ...)
	T:/gopath/src/github.com/docker/docker/api/client/hijack.go:242 +0x1147
github.com/docker/docker/api/client.(*DockerCli).CmdExec.func2(0x0, 0x0)
	T:/gopath/src/github.com/docker/docker/api/client/exec.go:95 +0x1ff
github.com/docker/docker/pkg/promise.Go.func1(0xc08200d980, 0xc08200d920)
	T:/gopath/src/github.com/docker/docker/pkg/promise/promise.go:8 +0x24
created by github.com/docker/docker/pkg/promise.Go
	T:/gopath/src/github.com/docker/docker/pkg/promise/promise.go:9 +0x64

goroutine 9 [IO wait]:
net.runtime_pollWait(0x1276018, 0x72, 0xc082157920)
	T:/Go/src/runtime/netpoll.go:157 +0x67
net.(*pollDesc).Wait(0xc08201b4f0, 0x72, 0x0, 0x0)
	T:/Go/src/net/fd_poll_runtime.go:73 +0x41
net.(*ioSrv).ExecIO(0xc082024188, 0xc08201b3e0, 0xa23138, 0x7, 0xb26870, 0x360000, 0x0, 0x0)
	T:/Go/src/net/fd_windows.go:182 +0x177
net.(*netFD).Read(0xc08201b380, 0xc08215f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	T:/Go/src/net/fd_windows.go:463 +0x17e
net.(*conn).Read(0xc082024198, 0xc08215f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	T:/Go/src/net/net.go:172 +0xeb
net/http.noteEOFReader.Read(0x12760d8, 0xc082024198, 0xc082098478, 0xc08215f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	T:/Go/src/net/http/transport.go:1370 +0x6e
net/http.(*noteEOFReader).Read(0xc082164420, 0xc08215f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd7
bufio.(*Reader).fill(0xc08200d740)
	T:/Go/src/bufio/bufio.go:97 +0x1f0
bufio.(*Reader).Peek(0xc08200d740, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	T:/Go/src/bufio/bufio.go:132 +0xd3
net/http.(*persistConn).readLoop(0xc082098420)
	T:/Go/src/net/http/transport.go:876 +0xfe
created by net/http.(*Transport).dialConn
	T:/Go/src/net/http/transport.go:685 +0xc7f

goroutine 10 [select]:
net/http.(*persistConn).writeLoop(0xc082098420)
	T:/Go/src/net/http/transport.go:1009 +0x413
created by net/http.(*Transport).dialConn
	T:/Go/src/net/http/transport.go:686 +0xca4

goroutine 12 [runnable]:
github.com/docker/docker/pkg/term.restoreAtInterrupt.func1(0xc08200daa0, 0x3, 0xc082157b80)
	T:/gopath/src/github.com/docker/docker/pkg/term/term_windows.go:200
created by github.com/docker/docker/pkg/term.restoreAtInterrupt
	T:/gopath/src/github.com/docker/docker/pkg/term/term_windows.go:204 +0xee

goroutine 13 [runnable]:
github.com/docker/docker/api/client.(*DockerCli).hijackWithContentType.func2(0x1274c00, 0xc08200d1a0, 0xc082024101, 0xc082072300, 0xc082157b80, 0x12742c0, 0xc082024010, 0xc0821579a0, 0xc08200da40, 0x12742c0, ...)
	T:/gopath/src/github.com/docker/docker/api/client/hijack.go:204
created by github.com/docker/docker/api/client.(*DockerCli).hijackWithContentType
	T:/gopath/src/github.com/docker/docker/api/client/hijack.go:222 +0x11e4
````

--- a/pkg/term/term_windows.go
+++ b/pkg/term/term_windows.go
@@ -40,8 +40,8 @@ var usingNativeConsole bool
 func StdStreams() (stdIn io.ReadCloser, stdOut, stdErr io.Writer) {
 	switch {
 	case os.Getenv("ConEmuANSI") == "ON":
-		// The ConEmu shell emulates ANSI well by default.
-		return os.Stdin, os.Stdout, os.Stderr
+		// The ConEmu terminal emulates ANSI on output streams well.
+		return windows.ConEmuStreams()
 	case os.Getenv("MSYSTEM") != "":
 		// MSYS (mingw) does not emulate ANSI well.
 		return windows.ConsoleStreams()

--- a/pkg/term/windows/console.go
+++ b/pkg/term/windows/console.go
@@ -8,7 +8,43 @@ import (
 	"syscall"
 
 	"github.com/Azure/go-ansiterm/winterm"
+
+	ansiterm "github.com/Azure/go-ansiterm"
+	"github.com/Sirupsen/logrus"
+	"io/ioutil"
 )
+
+// ConEmuStreams returns prepared versions of console streams,
+// for proper use in ConEmu terminal.
+// The ConEmu terminal emulates ANSI on output streams well by default.
+func ConEmuStreams() (stdIn io.ReadCloser, stdOut, stdErr io.Writer) {
+	if IsConsole(os.Stdin.Fd()) {
+		stdIn = newAnsiReader(syscall.STD_INPUT_HANDLE)
+	} else {
+		stdIn = os.Stdin
+	}
+
+	stdOut = os.Stdout
+	stdErr = os.Stderr
+
+	// WARNING (BEGIN): sourced from newAnsiWriter
+
+	logFile := ioutil.Discard
+
+	if isDebugEnv := os.Getenv(ansiterm.LogEnv); isDebugEnv == "1" {
+		logFile, _ = os.Create("ansiReaderWriter.log")
+	}
+
+	logger = &logrus.Logger{
+		Out:       logFile,
+		Formatter: new(logrus.TextFormatter),
+		Level:     logrus.DebugLevel,
+	}
+
+	// WARNING (END): sourced from newAnsiWriter
+
+	return stdIn, stdOut, stdErr
+}
 
 // ConsoleStreams returns a wrapped version for each standard stream referencing a console,
 // that handles ANSI character sequences.

